p2p/ipld/plugin/plugin_test.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"math/rand"
+	"sort"
+	"testing"
+
+	shell "github.com/ipfs/go-ipfs-api"
+	"github.com/lazyledger/nmt"
+)
+
+func TestDataSquareRowOrColumnRawInputParserCidEqNmtRoot(t *testing.T) {
+	tests := []struct {
+		name     string
+		leafData [][]byte
+	}{
+		{"16 leaves", generateRandNamespacedRawData(16, namespaceSize, shareSize)},
+		// TODO add at least a row of an extended data square (incl. parity bytes) as a test-vector too
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			n := nmt.New(sha256.New())
+			buf := createByteBufFromRawData(t, tt.leafData)
+			for _, share := range tt.leafData {
+				err := n.Push(share[:namespaceSize], share[namespaceSize:])
+				if err != nil {
+					t.Errorf("nmt.Push() unexpected error = %v", err)
+					return
+				}
+			}
+			gotNodes, err := DataSquareRowOrColumnRawInputParser(buf, 0, 0)
+			if err != nil {
+				t.Errorf("DataSquareRowOrColumnRawInputParser() unexpected error = %v", err)
+				return
+			}
+			rootNodeCid := gotNodes[0].Cid()
+			multiHashOverhead := 2
+			lastNodeHash := rootNodeCid.Hash()
+			if got, want := lastNodeHash[multiHashOverhead:], n.Root().Bytes(); !bytes.Equal(got, want) {
+				t.Errorf("hashes don't match\ngot: %v\nwant: %v", got, want)
+			}
+			lastNodeCid := gotNodes[len(gotNodes)-1].Cid()
+			if gotHash, wantHash := lastNodeCid.Hash(), hashLeaf(tt.leafData[0]); !bytes.Equal(gotHash[multiHashOverhead:], wantHash) {
+				t.Errorf("first node's hash does not match the Cid\ngot: %v\nwant: %v", gotHash[multiHashOverhead:], wantHash)
+			}
+			nodePrefixOffset := 1 // leaf / inner node prefix is one byte
+			lastLeafNodeData := gotNodes[len(gotNodes)-1].RawData()
+			if gotData, wantData := lastLeafNodeData[nodePrefixOffset:], tt.leafData[0]; !bytes.Equal(gotData, wantData) {
+				t.Errorf("first node's data does not match the leaf's data\ngot: %v\nwant: %v", gotData, wantData)
+			}
+		})
+	}
+}
+
+func TestDagPutWithPlugin(t *testing.T) {
+	//t.Skip("Requires running ipfs daemon with the plugin compiled and installed")
```

Comment:
Note to myself: after some more tests, this test should actually be skipped. Maybe it can be replaced by a test that runs and ipfs node and loads the plugin instead of relying on a running daemon.

```suggestion
	t.Skip("Requires running ipfs daemon with the plugin compiled and installed")
```